src/sqlfluff/utils/reflow/sequence.py

```diff
@@ -20,6 +21,112 @@
+            and "newline" not in elements[prev_nl_idx].class_types
+            and not any(seg.is_code for seg in elements[prev_nl_idx - 1].segments)
+        ):
+            prev_nl_idx -= 2
+        while (
+            next_nl_idx < len(elements) - 2
+            and "newline" not in elements[next_nl_idx].class_types
+            and not any(seg.is_code for seg in elements[next_nl_idx + 1].segments)
+        ):
+            next_nl_idx += 2
+        # Then just find the next code
+        prev_code_pt_idx = prev_nl_idx
+        next_code_pt_idx = next_nl_idx
+        # We hop in 2s because we're checking two ahead.
+        while next_code_pt_idx < len(elements) - 2 and not any(
+            seg.is_code for seg in elements[next_code_pt_idx + 1].segments
+        ):
+            next_code_pt_idx += 2
+        while prev_code_pt_idx >= 2 and not any(
+            seg.is_code for seg in elements[prev_code_pt_idx - 1].segments
+        ):
+            prev_code_pt_idx -= 2
+        return cls(
+            span.target,
+            prev_code_pt_idx,
+            prev_nl_idx,
+            prev_point_idx,
+            next_point_idx,
+            next_nl_idx,
+            next_code_pt_idx,
+            span.line_position,
+        )
+
+    def has_templated_newline(self, elements: ReflowSequenceType) -> bool:
+        """Is either side a templated newline?
+
+        If either side has a templated newline, then that's ok too.
+        The intent here is that if the next newline is a _templated_
+        one, then in the source there will be a tag ({{ tag }}), which
+        acts like _not having a newline_.
+        """
+        # Check the _last_ newline of the previous point.
+        # Slice backward to search in reverse.
+        for seg in elements[self.prev_nl_idx].segments[::-1]:
+            if seg.is_type("newline"):
+                if not seg.pos_marker.is_literal():
+                    return True
+                break
+        # Check the _first_ newline of the next point.
+        for seg in elements[self.next_nl_idx].segments:
+            if seg.is_type("newline"):
+                if not seg.pos_marker.is_literal():
+                    return True
+                break
+        return False
+
+    def has_inappropriate_newlines(self, elements: ReflowSequenceType) -> bool:
+        """Is the span surrounded by one (but not two) line breaks?"""
+        # Here we use the newline index, not
+        # just the adjacent point, so that we can see past comments.
+        n_prev_newlines = elements[self.prev_nl_idx].num_newlines()
+        n_next_newlines = elements[self.next_nl_idx].num_newlines()
+        return (
+            # If there isn't a newline on either side then carry
+            # on. We're not interested in this yet.
+            not bool(n_prev_newlines or n_next_newlines)
+            # If there is a newline on BOTH sides. That's ok.
+            or bool(n_prev_newlines and n_next_newlines)
```

Comment:
For readability, could we rewrite this using comparison operators? I find this a bit tricky to read.

Reply:
Check out the new version - is it a bit clearer?